tests/python/test_ad_dynamic_index.py

```diff
@@ -0,0 +1,28 @@
+import taichi as ti
+from tests import test_utils
+
+
+@test_utils.test(require=ti.extension.dynamic_index,
```

Comment:
nit: since you also handled ptroffset in MakeDual, maybe it's good to add another fwd mode test with dynamic index=true?

Reply:
Sure. The current API of fwd mode doesn't support vector or matrix field. I can submit another PR to enable vector or matrix field in fwd mode and attach a fwd mode test.